diff -Naur nova-2011.2.orig/nova/virt/libvirt_conn.py nova-2011.2/nova/virt/libvirt_conn.py
--- nova-2011.2.orig/nova/virt/libvirt_conn.py	2011-03-15 03:35:15.000000000 +0300
+++ nova-2011.2/nova/virt/libvirt_conn.py	2011-03-15 18:19:38.341597125 +0300
@@ -984,7 +984,7 @@
 
         xml = self._conn.getCapabilities()
         xml = libxml2.parseDoc(xml)
-        nodes = xml.xpathEval('//cpu')
+        nodes = xml.xpathEval('/capabilities/host/cpu')
         if len(nodes) != 1:
             raise exception.Invalid(_("Invalid xml. '<cpu>' must be 1,"
                                       "but %d\n") % len(nodes)
