*** nova/virt/libvirt/connection.py.orig	2011-08-12 21:46:19.000000000 +0300
--- nova/virt/libvirt/connection.py	2011-08-12 22:00:40.870712653 +0300
***************
*** 685,691 ****
          ajaxterm_cmd = 'sudo socat - %s' \
                         % get_pty_for_instance(instance['name'])
  
!         cmd = ['%s/tools/ajaxterm/ajaxterm.py' % utils.novadir(),
                 '--command', ajaxterm_cmd, '-t', token, '-p', port]
  
          utils.execute(cmd)
--- 685,691 ----
          ajaxterm_cmd = 'sudo socat - %s' \
                         % get_pty_for_instance(instance['name'])
  
!         cmd = ['/usr/share/nova/ajaxterm/ajaxterm.py',
                 '--command', ajaxterm_cmd, '-t', token, '-p', port]
  
          utils.execute(cmd)
