=== modified file 'tools/euca-get-ajax-console'
--- tools/euca-get-ajax-console	2011-03-18 13:56:05 +0000
+++ tools/euca-get-ajax-console	2011-04-01 15:11:13 +0000
@@ -35,6 +35,7 @@
 import boto
 import nova
 from boto.ec2.connection import EC2Connection
+import euca2ools
 from euca2ools import Euca2ool, InstanceValidationError, Util
 
 usage_string = """
@@ -93,8 +94,13 @@
                              aws_secret_access_key, **kwargs)
 
 # override boto's connect_ec2 method, so that we can use NovaEC2Connection
+# (This is for Euca2ools 1.2)
 boto.connect_ec2 = override_connect_ec2
 
+# Override Euca2ools' EC2Connection class (which it gets from boto)
+# (This is for Euca2ools 1.3)
+euca2ools.EC2Connection = NovaEC2Connection
+
 
 def usage(status=1):
     print usage_string
