diff -Naur openstack-dashboard-1.0.orig/dashboard/settings.py openstack-dashboard-1.0/dashboard/settings.py
--- openstack-dashboard-1.0.orig/dashboard/settings.py	2011-07-04 08:12:00.000000000 +0400
+++ openstack-dashboard-1.0/dashboard/settings.py	2011-07-06 21:18:38.090433478 +0400
@@ -116,6 +116,9 @@
 
 TOTAL_CLOUD_RAM_GB = 10
 
+SYSCONFDIR = '/etc'
+sys.path.insert(0, SYSCONFDIR + '/dashboard')
+
 try:
     from local.local_settings import *
 except Exception, e:
diff -Naur openstack-dashboard-1.0.orig/local/local_settings.py.example openstack-dashboard-1.0/local/local_settings.py.example
--- openstack-dashboard-1.0.orig/local/local_settings.py.example	2011-07-04 08:12:00.000000000 +0400
+++ openstack-dashboard-1.0/local/local_settings.py.example	2011-07-06 22:04:20.539178375 +0400
@@ -9,7 +9,7 @@
 DATABASES = {
     'default': {
         'ENGINE': 'django.db.backends.sqlite3',
-        'NAME': os.path.join(LOCAL_PATH, 'dashboard_openstack.sqlite3'),
+        'NAME': '/var/lib/dashboard/dashboard_openstack.sqlite3',
     },
 }
