diff -Naur nova-2011.1.orig/bin/nova-api nova-2011.1/bin/nova-api
--- nova-2011.1.orig/bin/nova-api	2011-01-17 22:22:52.000000000 +0300
+++ nova-2011.1/bin/nova-api	2011-01-18 13:33:47.387946635 +0300
@@ -79,8 +79,8 @@
 
 if __name__ == '__main__':
     FLAGS(sys.argv)
-    conf = wsgi.paste_config_file('nova-api.conf')
+    conf = wsgi.paste_config_file('/etc/nova/nova-api.conf')
     if conf:
         run_app(conf)
     else:
-        LOG.error(_("No paste configuration found for: %s"), 'nova-api.conf')
+        LOG.error(_("No paste configuration found for: %s"), '/etc/nova/nova-api.conf')
diff -Naur nova-2011.1.orig/nova/utils.py nova-2011.1/nova/utils.py
--- nova-2011.1.orig/nova/utils.py	2011-01-17 22:22:52.000000000 +0300
+++ nova-2011.1/nova/utils.py	2011-01-18 13:34:06.894921472 +0300
@@ -161,7 +161,7 @@
     return os.path.abspath(nova.__file__).split('nova/__init__.pyc')[0]
 
 
-def default_flagfile(filename='nova.conf'):
+def default_flagfile(filename='/etc/nova/nova.conf'):
     for arg in sys.argv:
         if arg.find('flagfile') != -1:
             break
