=== modified file 'nova/virt/disk.py'
--- nova/virt/disk.py	2011-05-12 20:42:04 +0000
+++ nova/virt/disk.py	2011-09-06 11:29:50 +0000
@@ -2,6 +2,9 @@
 
 # Copyright 2010 United States Government as represented by the
 # Administrator of the National Aeronautics and Space Administration.
+#
+# Copyright 2011 Grid Dynamics
+#
 # All Rights Reserved.
 #
 #    Licensed under the Apache License, Version 2.0 (the "License"); you may
@@ -20,8 +23,14 @@
 
 Includes injection of SSH PGP keys into authorized_keys file.
 
+**Related Flags**
+
+injected_network_template_dir:  Template directory for injected network
+inject_ssh_key:  Enable or disable ssh key injection
+
 """
 
+import json
 import os
 import tempfile
 import time
@@ -33,6 +42,9 @@
 from nova import log as logging
 from nova import utils
 
+from Cheetah.Template import Template
+import guestfs
+
 
 LOG = logging.getLogger('nova.compute.disk')
 FLAGS = flags.FLAGS
@@ -40,14 +52,16 @@
                      'minimum size in bytes of root partition')
 flags.DEFINE_integer('block_size', 1024 * 1024 * 256,
                      'block_size to use for dd')
-flags.DEFINE_string('injected_network_template',
+flags.DEFINE_string('injected_network_template_dir',
                     utils.abspath('virt/interfaces.template'),
                     'Template file for injected network')
-flags.DEFINE_integer('timeout_nbd', 10,
-                     'time to wait for a NBD device coming up')
-flags.DEFINE_integer('max_nbd_devices', 16,
-                     'maximum number of possible nbd devices')
+flags.DEFINE_bool('inject_ssh_key', True, 'inject ssh key')
 
+path_map = {}
+path_map['ubuntu'] = '/etc/networking/interfaces'
+path_map['rhel'] = '/etc/sysconfig/network-scripts/ifcfg-$name'
+# Note (Ivan K.): use RHEL config as default
+path_map['unknown'] = '/etc/sysconfig/network-scripts/ifcfg-$name'
 
 def extend(image, size):
     """Increase image to size"""
@@ -60,7 +74,7 @@
     utils.execute('resize2fs', image, check_exit_code=False)
 
 
-def inject_data(image, key=None, net=None, partition=None, nbd=False):
+def inject_data(image, key=None, nets=[], partition=None, nbd=False):
     """Injects a ssh key and optionally net data into a disk image.
 
     it will mount the image as a fully partitioned disk and attempt to inject
@@ -69,173 +83,88 @@
     If partition is not specified it mounts the image as a single partition.
 
     """
-    device = _link_device(image, nbd)
-    try:
-        if not partition is None:
-            # create partition
-            out, err = utils.execute('sudo', 'kpartx', '-a', device)
-            if err:
-                raise exception.Error(_('Failed to load partition: %s') % err)
-            mapped_device = '/dev/mapper/%sp%s' % (device.split('/')[-1],
-                                                   partition)
-        else:
-            mapped_device = device
-
-        try:
-            # We can only loopback mount raw images. If the device isn't there,
-            # it's normally because it's a .vmdk or a .vdi etc
-            if not os.path.exists(mapped_device):
-                raise exception.Error('Mapped device was not found (we can'
-                                      ' only inject raw disk images): %s' %
-                                      mapped_device)
-
-            # Configure ext2fs so that it doesn't auto-check every N boots
-            out, err = utils.execute('sudo', 'tune2fs',
-                                     '-c', 0, '-i', 0, mapped_device)
-
-            tmpdir = tempfile.mkdtemp()
-            try:
-                # mount loopback to dir
-                out, err = utils.execute(
-                        'sudo', 'mount', mapped_device, tmpdir)
-                if err:
-                    raise exception.Error(_('Failed to mount filesystem: %s')
-                                          % err)
-
-                try:
-                    inject_data_into_fs(tmpdir, key, net, utils.execute)
-                finally:
-                    # unmount device
-                    utils.execute('sudo', 'umount', mapped_device)
-            finally:
-                # remove temporary directory
-                utils.execute('rmdir', tmpdir)
-        finally:
-            if not partition is None:
-                # remove partitions
-                utils.execute('sudo', 'kpartx', '-d', device)
-    finally:
-        _unlink_device(device, nbd)
-
-
-def setup_container(image, container_dir=None, nbd=False):
-    """Setup the LXC container.
-
-    It will mount the loopback image to the container directory in order
-    to create the root filesystem for the container.
-
-    LXC does not support qcow2 images yet.
-    """
-    try:
-        device = _link_device(image, nbd)
-        utils.execute('sudo', 'mount', device, container_dir)
-    except Exception, exn:
-        LOG.exception(_('Failed to mount filesystem: %s'), exn)
-        _unlink_device(device, nbd)
-
-
-def destroy_container(target, instance, nbd=False):
-    """Destroy the container once it terminates.
-
-    It will umount the container that is mounted, try to find the loopback
-    device associated with the container and delete it.
-
-    LXC does not support qcow2 images yet.
-    """
-    try:
-        container_dir = '%s/rootfs' % target
-        utils.execute('sudo', 'umount', container_dir)
-    finally:
-        out, err = utils.execute('sudo', 'losetup', '-a')
-        for loop in out.splitlines():
-            if instance['name'] in loop:
-                device = loop.split(loop, ':')
-                _unlink_device(device, nbd)
-
-
-def _link_device(image, nbd):
-    """Link image to device using loopback or nbd"""
-    if nbd:
-        device = _allocate_device()
-        utils.execute('sudo', 'qemu-nbd', '-c', device, image)
-        # NOTE(vish): this forks into another process, so give it a chance
-        #             to set up before continuuing
-        for i in xrange(FLAGS.timeout_nbd):
-            if os.path.exists("/sys/block/%s/pid" % os.path.basename(device)):
-                return device
-            time.sleep(1)
-        raise exception.Error(_('nbd device %s did not show up') % device)
-    else:
-        out, err = utils.execute('sudo', 'losetup', '--find', '--show', image)
-        if err:
-            raise exception.Error(_('Could not attach image to loopback: %s')
-                                  % err)
-        return out.strip()
-
-
-def _unlink_device(device, nbd):
-    """Unlink image from device using loopback or nbd"""
-    if nbd:
-        utils.execute('sudo', 'qemu-nbd', '-d', device)
-        _free_device(device)
-    else:
-        utils.execute('sudo', 'losetup', '--detach', device)
-
-
-_DEVICES = ['/dev/nbd%s' % i for i in xrange(FLAGS.max_nbd_devices)]
-
-
-def _allocate_device():
-    # NOTE(vish): This assumes no other processes are allocating nbd devices.
-    #             It may race cause a race condition if multiple
-    #             workers are running on a given machine.
-    while True:
-        if not _DEVICES:
-            raise exception.Error(_('No free nbd devices'))
-        device = _DEVICES.pop()
-        if not os.path.exists("/sys/block/%s/pid" % os.path.basename(device)):
-            break
-    return device
-
-
-def _free_device(device):
-    _DEVICES.append(device)
-
-
-def inject_data_into_fs(fs, key, net, execute):
-    """Injects data into a filesystem already mounted by the caller.
-    Virt connections can call this directly if they mount their fs
-    in a different way to inject_data
-    """
-    if key:
-        _inject_key_into_fs(key, fs, execute=execute)
-    if net:
-        _inject_net_into_fs(net, fs, execute=execute)
-
-
-def _inject_key_into_fs(key, fs, execute=None):
-    """Add the given public ssh key to root's authorized_keys.
-
-    key is an ssh key string.
-    fs is the path to the base of the filesystem into which to inject the key.
-    """
-    sshdir = os.path.join(fs, 'root', '.ssh')
-    utils.execute('sudo', 'mkdir', '-p', sshdir)  # existing dir doesn't matter
-    utils.execute('sudo', 'chown', 'root', sshdir)
-    utils.execute('sudo', 'chmod', '700', sshdir)
-    keyfile = os.path.join(sshdir, 'authorized_keys')
-    utils.execute('sudo', 'tee', '-a', keyfile,
-            process_input='\n' + key.strip() + '\n')
-
-
-def _inject_net_into_fs(net, fs, execute=None):
-    """Inject /etc/network/interfaces into the filesystem rooted at fs.
-
-    net is the contents of /etc/network/interfaces.
-    """
-    netdir = os.path.join(os.path.join(fs, 'etc'), 'network')
-    utils.execute('sudo', 'mkdir', '-p', netdir)  # existing dir doesn't matter
-    utils.execute('sudo', 'chown', 'root:root', netdir)
-    utils.execute('sudo', 'chmod', 755, netdir)
-    netfile = os.path.join(netdir, 'interfaces')
-    utils.execute('sudo', 'tee', netfile, process_input=net)
+    g = guestfs.GuestFS()
+    try:
+        LOG.debug(_("Mounting disk..."))
+        g.add_drive(image)
+        g.launch()
+        devices = g.inspect_os()
+        device = devices[0]
+        g.mount(device,"/")
+        if key and FLAGS.inject_ssh_key:
+            LOG.debug(_("Injecting SSH key..."))
+            sshdir = "/root/.ssh"
+            path = sshdir + "/authorized_keys"
+            if not g.is_dir(sshdir):
+                g.mkdir_p(sshdir)
+                g.chmod(0700, sshdir)
+            g.write(path, key)
+            g.chmod(0600, path)
+
+        net = []
+        os_name = g.inspect_get_distro(device)
+
+        if os_name == "ubuntu":
+            ifc_template = open(os.path.join(FLAGS.injected_network_template_dir, "interfaces.ubuntu.template")).read()
+            net = [
+                {'name': x['name'],
+                 'data':
+                     str
+                         (Template(ifc_template,
+                                  searchList=[x, {'use_ipv6': FLAGS.use_ipv6}]))}
+                    for x in nets]
+            if len(net):
+                net = str(Template(ifc_template,
+                               searchList=[{'interfaces': nets,
+                                            'use_ipv6': FLAGS.use_ipv6}]))
+                LOG.debug(_("Injecting network for interface %(name)s..."
+                                                                        % net))
+                path = get_network_interface_config_path(g, device, net)
+                if path:
+                    g.write(path, net['data'])
+                else:
+                    raise Exception
+        if os_name == "rhel":
+            ifc_template = open(os.path.join(FLAGS.injected_network_template_dir, "interfaces.rhel.template")).read()
+            net = [
+                {'name': x['name'],
+                 'data':
+                      str
+                       (Template(ifc_template,
+                             searchList=[x, {'use_ipv6': FLAGS.use_ipv6}]))}
+                     for x in nets]
+            if len(net):
+                for n in net:
+                     LOG.debug(_("Injecting network for interface %(name)s..."
+                                                                         % n))
+                     path = get_network_interface_config_path(g, device, n)
+                     g.write(path, n['data'])
+    except Exception as e:
+        LOG.exception(e)
+        raise
+    finally:
+        LOG.debug(_("Deleting guestfs object..."))
+        g.umount_all()
+        del(g)
+
+
+def get_network_interface_config_path(gfs, device, net):
+    """ Returns path to the file with network interfaces configuration
+
+    It will return linux distributive specific path for
+    the network interface configuration file.
+
+    Returns path to RHEL configuration file by default
+    """
+
+    distro = gfs.inspect_get_distro(device)
+    path = str(Template(path_map[distro], searchList=[{'name': net['name']}]))
+    if not gfs.is_file(path):
+        for template in path_map.itervalues():
+            p = str(Template(template, searchList=[{'name': net['name']}]))
+            if gfs.is_file(p):
+                path = p
+                break
+    if path != 'unknown':
+        return path
+    raise IOError('Can not get path to network interfaces config.')

=== modified file 'nova/virt/libvirt/connection.py'
--- nova/virt/libvirt/connection.py	2011-06-03 21:43:12 +0000
+++ nova/virt/libvirt/connection.py	2011-09-06 11:29:56 +0000
@@ -31,7 +31,6 @@
 :rescue_image_id:  Rescue ami image (default: ami-rescue).
 :rescue_kernel_id:  Rescue aki image (default: aki-rescue).
 :rescue_ramdisk_id:  Rescue ari image (default: ari-rescue).
-:injected_network_template:  Template file for injected network
 :allow_project_net_traffic:  Whether to allow in project network traffic
 
 """
@@ -855,7 +854,6 @@
         net = None
 
         nets = []
-        ifc_template = open(FLAGS.injected_network_template).read()
         ifc_num = -1
         have_injected_networks = False
         admin_context = context.get_admin_context()
@@ -881,22 +879,17 @@
                    'netmask_v6': network_ref['netmask_v6']}
             nets.append(net_info)
 
-        if have_injected_networks:
-            net = str(Template(ifc_template,
-                               searchList=[{'interfaces': nets,
-                                            'use_ipv6': FLAGS.use_ipv6}]))
-
-        if key or net:
+        if key or len(nets):
             inst_name = inst['name']
             img_id = inst.image_ref
             if key:
                 LOG.info(_('instance %(inst_name)s: injecting key into'
                         ' image %(img_id)s') % locals())
-            if net:
+            if len(nets):
                 LOG.info(_('instance %(inst_name)s: injecting net into'
                         ' image %(img_id)s') % locals())
             try:
-                disk.inject_data(basepath('disk'), key, net,
+                disk.inject_data(basepath('disk'), key, nets,
                                  partition=target_partition,
                                  nbd=FLAGS.use_cow_images)
 

