diff -Naur nova-2011.1.orig/nova/utils.py nova-2011.1/nova/utils.py
--- nova-2011.1.orig/nova/utils.py	2011-01-21 19:30:29.000000000 +0300
+++ nova-2011.1/nova/utils.py	2011-01-21 19:46:40.403785449 +0300
@@ -161,7 +161,7 @@
     return os.path.abspath(nova.__file__).split('nova/__init__.pyc')[0]
 
 
-def default_flagfile(filename='nova.conf'):
+def default_flagfile(filename='/etc/nova/nova.conf'):
     for arg in sys.argv:
         if arg.find('flagfile') != -1:
             break
diff -Naur nova-2011.2.orig/nova/service.py nova-2011.2/nova/service.py
--- nova-2011.2.orig/nova/service.py	2011-03-13 19:19:49.000000000 +0300
+++ nova-2011.2/nova/service.py	2011-03-14 15:59:06.798683246 +0300
@@ -56,7 +56,7 @@
 flags.DEFINE_string('osapi_listen', "0.0.0.0",
                     'IP address for OpenStack API to listen')
 flags.DEFINE_integer('osapi_listen_port', 8774, 'port for os api to listen')
-flags.DEFINE_string('api_paste_config', "api-paste.ini",
+flags.DEFINE_string('api_paste_config', "/etc/nova/api-paste.ini",
                     'File name for the paste.deploy config for nova-api')
 
 
