diff -Naur nova-2011.2.orig/nova/CA/openssl.cnf.tmpl nova-2011.2/nova/CA/openssl.cnf.tmpl
--- nova-2011.2.orig/nova/CA/openssl.cnf.tmpl	2011-04-06 11:02:43.000000000 +0400
+++ nova-2011.2/nova/CA/openssl.cnf.tmpl	2011-04-06 16:50:03.243191642 +0400
@@ -43,7 +43,7 @@
 
 [ policy_match ]
 countryName		= match
-stateOrProvinceName	= match
+stateOrProvinceName	= supplied
 organizationName	= optional
 organizationalUnitName	= optional
 commonName		= supplied
