=== modified file 'nova/virt/disk.py'
--- nova/virt/disk.py	2011-05-12 20:42:04 +0000
+++ nova/virt/disk.py	2011-07-26 09:11:56 +0000
@@ -4,6 +4,8 @@
 # Administrator of the National Aeronautics and Space Administration.
 # All Rights Reserved.
 #
+# Copyright 2011 Grid Dynamics
+#
 #    Licensed under the Apache License, Version 2.0 (the "License"); you may
 #    not use this file except in compliance with the License. You may obtain
 #    a copy of the License at
@@ -20,6 +22,10 @@
 
 Includes injection of SSH PGP keys into authorized_keys file.
 
+**Related Flags**
+
+:injected_network_template:  Template file for injected network
+
 """
 
 import os
@@ -33,6 +39,9 @@
 from nova import log as logging
 from nova import utils
 
+from Cheetah.Template import Template
+import guestfs
+
 
 LOG = logging.getLogger('nova.compute.disk')
 FLAGS = flags.FLAGS
@@ -43,11 +52,12 @@
 flags.DEFINE_string('injected_network_template',
                     utils.abspath('virt/interfaces.template'),
                     'Template file for injected network')
-flags.DEFINE_integer('timeout_nbd', 10,
-                     'time to wait for a NBD device coming up')
-flags.DEFINE_integer('max_nbd_devices', 16,
-                     'maximum number of possible nbd devices')
 
+path_map = {}
+path_map['ubuntu'] = '/etc/networking/interfaces'
+path_map['rhel'] = '/etc/sysconfig/network-scripts/ifcfg-$name'
+# Note (Ivan K.): use RHEL config as default
+path_map['unknown'] = '/etc/sysconfig/network-scripts/ifcfg-$name'
 
 def extend(image, size):
     """Increase image to size"""
@@ -60,62 +70,72 @@
     utils.execute('resize2fs', image, check_exit_code=False)
 
 
-def inject_data(image, key=None, net=None, partition=None, nbd=False):
+def inject_data(image, key=None, nets=[], partition=None, nbd=False):
     """Injects a ssh key and optionally net data into a disk image.
-
-    it will mount the image as a fully partitioned disk and attempt to inject
-    into the specified partition number.
-
-    If partition is not specified it mounts the image as a single partition.
-
     """
-    device = _link_device(image, nbd)
+    g = guestfs.GuestFS()
     try:
-        if not partition is None:
-            # create partition
-            out, err = utils.execute('sudo', 'kpartx', '-a', device)
-            if err:
-                raise exception.Error(_('Failed to load partition: %s') % err)
-            mapped_device = '/dev/mapper/%sp%s' % (device.split('/')[-1],
-                                                   partition)
-        else:
-            mapped_device = device
-
-        try:
-            # We can only loopback mount raw images. If the device isn't there,
-            # it's normally because it's a .vmdk or a .vdi etc
-            if not os.path.exists(mapped_device):
-                raise exception.Error('Mapped device was not found (we can'
-                                      ' only inject raw disk images): %s' %
-                                      mapped_device)
-
-            # Configure ext2fs so that it doesn't auto-check every N boots
-            out, err = utils.execute('sudo', 'tune2fs',
-                                     '-c', 0, '-i', 0, mapped_device)
-
-            tmpdir = tempfile.mkdtemp()
-            try:
-                # mount loopback to dir
-                out, err = utils.execute(
-                        'sudo', 'mount', mapped_device, tmpdir)
-                if err:
-                    raise exception.Error(_('Failed to mount filesystem: %s')
-                                          % err)
-
-                try:
-                    inject_data_into_fs(tmpdir, key, net, utils.execute)
-                finally:
-                    # unmount device
-                    utils.execute('sudo', 'umount', mapped_device)
-            finally:
-                # remove temporary directory
-                utils.execute('rmdir', tmpdir)
-        finally:
-            if not partition is None:
-                # remove partitions
-                utils.execute('sudo', 'kpartx', '-d', device)
+        LOG.debug(_("Mounting disk..."))
+        g.add_drive(image)
+        g.launch()
+        devices = g.inspect_os()
+        device = devices[0]
+        g.mount(device,"/")
+        if key:
+            LOG.debug(_("Injecting SSH key..."))
+            sshdir = "/root/.ssh"
+            path = sshdir + "/authorized_keys"
+            if not g.is_dir(sshdir):
+                g.mkdir_p(sshdir)
+                g.chmod(0700, sshdir)
+            g.write(path, key)
+            g.chmod(0600, path)
+
+        net = []
+        os_name = get_os_name(g, device, None)
+
+        if os_name == "ubuntu":
+            ifc_template = open(os.path.join(FLAGS.injected_network_template, "interfaces", ".template")).read()
+            net = [
+                {'name': x['name'],
+                 'data':
+                     str
+                         (Template(ifc_template,
+                                  searchList=[x, {'use_ipv6': FLAGS.use_ipv6}]))}
+                    for x in nets]
+            if len(net):
+                net = str(Template(ifc_template,
+                               searchList=[{'interfaces': nets,
+                                            'use_ipv6': FLAGS.use_ipv6}]))
+                LOG.debug(_("Injecting network for interface %(name)s..."
+                                                                        % net))
+                path = get_network_interface_config_path(g, device, net)
+                if path:
+                    g.write(path, net['data'])
+                else:
+                    raise Exception
+        if os_name == "rhel":
+            ifc_template = open(os.path.join(FLAGS.injected_network_template, "interfaces.rhel", ".template")).read()
+            net = [
+                {'name': x['name'],
+                 'data':
+                     str
+                         (Template(ifc_template,
+                                  searchList=[x, {'use_ipv6': FLAGS.use_ipv6}]))}
+                    for x in nets]
+            if len(net):
+                for net in net:
+                    LOG.debug(_("Injecting network for interface %(name)s..."
+                                                                            % net))
+                    path = get_network_interface_config_path(g, device, net)
+                    g.write(path, net['data'])
+    except Exception as e:
+        LOG.exception(e)
+        raise
     finally:
-        _unlink_device(device, nbd)
+        LOG.debug(_("Deleting guestfs object..."))
+        g.umount_all()
+        del(g)
 
 
 def setup_container(image, container_dir=None, nbd=False):
@@ -153,89 +173,36 @@
                 _unlink_device(device, nbd)
 
 
-def _link_device(image, nbd):
-    """Link image to device using loopback or nbd"""
-    if nbd:
-        device = _allocate_device()
-        utils.execute('sudo', 'qemu-nbd', '-c', device, image)
-        # NOTE(vish): this forks into another process, so give it a chance
-        #             to set up before continuuing
-        for i in xrange(FLAGS.timeout_nbd):
-            if os.path.exists("/sys/block/%s/pid" % os.path.basename(device)):
-                return device
-            time.sleep(1)
-        raise exception.Error(_('nbd device %s did not show up') % device)
-    else:
-        out, err = utils.execute('sudo', 'losetup', '--find', '--show', image)
-        if err:
-            raise exception.Error(_('Could not attach image to loopback: %s')
-                                  % err)
-        return out.strip()
-
-
-def _unlink_device(device, nbd):
-    """Unlink image from device using loopback or nbd"""
-    if nbd:
-        utils.execute('sudo', 'qemu-nbd', '-d', device)
-        _free_device(device)
-    else:
-        utils.execute('sudo', 'losetup', '--detach', device)
-
-
-_DEVICES = ['/dev/nbd%s' % i for i in xrange(FLAGS.max_nbd_devices)]
-
-
-def _allocate_device():
-    # NOTE(vish): This assumes no other processes are allocating nbd devices.
-    #             It may race cause a race condition if multiple
-    #             workers are running on a given machine.
-    while True:
-        if not _DEVICES:
-            raise exception.Error(_('No free nbd devices'))
-        device = _DEVICES.pop()
-        if not os.path.exists("/sys/block/%s/pid" % os.path.basename(device)):
-            break
-    return device
-
-
-def _free_device(device):
-    _DEVICES.append(device)
-
-
-def inject_data_into_fs(fs, key, net, execute):
-    """Injects data into a filesystem already mounted by the caller.
-    Virt connections can call this directly if they mount their fs
-    in a different way to inject_data
-    """
-    if key:
-        _inject_key_into_fs(key, fs, execute=execute)
-    if net:
-        _inject_net_into_fs(net, fs, execute=execute)
-
-
-def _inject_key_into_fs(key, fs, execute=None):
-    """Add the given public ssh key to root's authorized_keys.
-
-    key is an ssh key string.
-    fs is the path to the base of the filesystem into which to inject the key.
-    """
-    sshdir = os.path.join(fs, 'root', '.ssh')
-    utils.execute('sudo', 'mkdir', '-p', sshdir)  # existing dir doesn't matter
-    utils.execute('sudo', 'chown', 'root', sshdir)
-    utils.execute('sudo', 'chmod', '700', sshdir)
-    keyfile = os.path.join(sshdir, 'authorized_keys')
-    utils.execute('sudo', 'tee', '-a', keyfile,
-            process_input='\n' + key.strip() + '\n')
-
-
-def _inject_net_into_fs(net, fs, execute=None):
-    """Inject /etc/network/interfaces into the filesystem rooted at fs.
-
-    net is the contents of /etc/network/interfaces.
-    """
-    netdir = os.path.join(os.path.join(fs, 'etc'), 'network')
-    utils.execute('sudo', 'mkdir', '-p', netdir)  # existing dir doesn't matter
-    utils.execute('sudo', 'chown', 'root:root', netdir)
-    utils.execute('sudo', 'chmod', 755, netdir)
-    netfile = os.path.join(netdir, 'interfaces')
-    utils.execute('sudo', 'tee', netfile, process_input=net)
+def get_network_interface_config_path(gfs, device, net):
+    """ Returns path to the file with network interfaces configuration
+
+    It will return linux distributive specific path for
+    the network interface configuration file.
+
+    Returns path to RHEL configuration file by default
+    """
+
+    distro = gfs.inspect_get_distro(device)
+    path = str(Template(path_map[distro], searchList=[{'name': net}]))
+    if not gfs.is_file(path):
+        for template in path_map.itervalues():
+            p = str(Template(template, searchList=[{'name': net}]))
+            if gfs.is_file(p):
+                path = p
+                break
+    if path != 'unknown':
+        return path
+    raise IOError('Can not get path to network interfaces config.')
+
+
+def get_os_name(gfs, device, net):
+    distro = gfs.inspect_get_distro(device)
+    path = str(Template(path_map[distro], searchList=[{'name': net}]))
+    if not gfs.is_file(path):
+        for d, template in path_map.iteritems():
+            p = str(Template(template, searchList=[{'name': net}]))
+            if gfs.is_file(p):
+                distro =  d
+                break
+    return distro
+

=== modified file 'nova/virt/libvirt/connection.py'
--- nova/virt/libvirt/connection.py	2011-07-26 00:41:55 +0000
+++ nova/virt/libvirt/connection.py	2011-07-26 09:14:12 +0000
@@ -31,7 +31,6 @@
 :rescue_image_id:  Rescue ami image (default: ami-rescue).
 :rescue_kernel_id:  Rescue aki image (default: aki-rescue).
 :rescue_ramdisk_id:  Rescue ari image (default: ari-rescue).
-:injected_network_template:  Template file for injected network
 :allow_project_net_traffic:  Whether to allow in project network traffic
 
 """
@@ -896,12 +895,9 @@
             address = mapping['ips'][0]['ip']
             netmask = mapping['ips'][0]['netmask']
             address_v6 = None
-            gateway_v6 = None
-            netmask_v6 = None
             if FLAGS.use_ipv6:
                 address_v6 = mapping['ip6s'][0]['ip']
                 netmask_v6 = mapping['ip6s'][0]['netmask']
-                gateway_v6 = mapping['gateway6']
             net_info = {'name': 'eth%d' % ifc_num,
                    'address': address,
                    'netmask': netmask,
@@ -909,26 +905,21 @@
                    'broadcast': mapping['broadcast'],
                    'dns': mapping['dns'],
                    'address_v6': address_v6,
-                   'gateway6': gateway_v6,
+                   'gateway6': mapping['gateway6'],
                    'netmask_v6': netmask_v6}
             nets.append(net_info)
 
-        if have_injected_networks:
-            net = str(Template(ifc_template,
-                               searchList=[{'interfaces': nets,
-                                            'use_ipv6': FLAGS.use_ipv6}]))
-
-        if key or net:
+        if key or len(nets):
             inst_name = inst['name']
             img_id = inst.image_ref
             if key:
                 LOG.info(_('instance %(inst_name)s: injecting key into'
                         ' image %(img_id)s') % locals())
-            if net:
+            if len(nets):
                 LOG.info(_('instance %(inst_name)s: injecting net into'
                         ' image %(img_id)s') % locals())
             try:
-                disk.inject_data(basepath('disk'), key, net,
+                disk.inject_data(basepath('disk'), key, nets,
                                  partition=target_partition,
                                  nbd=FLAGS.use_cow_images)
 

