=== modified file 'nova/network/linux_net.py'
--- nova/network/linux_net.py	2011-05-11 19:24:01 +0000
+++ nova/network/linux_net.py	2011-05-20 13:27:24 +0000
@@ -27,6 +27,7 @@
 from nova import flags
 from nova import log as logging
 from nova import utils
+from IPy import IP
 
 
 LOG = logging.getLogger("nova.linux_net")
@@ -698,6 +699,7 @@
            '--listen-address=%s' % net['gateway'],
            '--except-interface=lo',
            '--dhcp-range=%s,static,120s' % net['dhcp_start'],
+           '--dhcp-lease-max=%s' % IP(net['cidr']).len(),
            '--dhcp-hostsfile=%s' % _dhcp_file(net['bridge'], 'conf'),
            '--dhcp-script=%s' % FLAGS.dhcpbridge,
            '--leasefile-ro']

