diff -Naur nova-2011.1.orig/nova/virt/libvirt.xml.template nova-2011.1/nova/virt/libvirt.xml.template
--- nova-2011.1.orig/nova/virt/libvirt.xml.template	2011-02-02 14:24:04.000000000 +0300
+++ nova-2011.1/nova/virt/libvirt.xml.template	2011-02-15 13:52:52.439935063 +0300
@@ -28,7 +28,7 @@
             #if $type == 'xen'
                 <cmdline>ro</cmdline>
             #else
-                <cmdline>root=/dev/vda console=ttyS0</cmdline>
+                <cmdline>root=/dev/vda console=ttyS0 selinux=0</cmdline>
             #end if
             #if $getVar('ramdisk', None)
                 <initrd>${ramdisk}</initrd>
