=== modified file 'nova/virt/libvirt/connection.py'
--- nova/virt/libvirt/connection.py	2011-05-19 20:25:57 +0000
+++ nova/virt/libvirt/connection.py	2011-05-27 14:59:20 +0000
@@ -669,8 +669,8 @@
         ajaxterm_cmd = 'sudo socat - %s' \
                        % get_pty_for_instance(instance['name'])
 
-        cmd = '%s/tools/ajaxterm/ajaxterm.py --command "%s" -t %s -p %s' \
-              % (utils.novadir(), ajaxterm_cmd, token, port)
+        cmd = '/usr/share/nova/ajaxterm/ajaxterm.py --command "%s" -t %s -p %s' \
+              % (ajaxterm_cmd, token, port)
 
         subprocess.Popen(cmd, shell=True)
         return {'token': token, 'host': host, 'port': port}

