=== modified file 'nova/virt/libvirt/connection.py'
--- nova/virt/libvirt/connection.py	2011-05-19 20:25:57 +0000
+++ nova/virt/libvirt/connection.py	2011-05-27 14:54:11 +0000
@@ -645,7 +645,7 @@
                 port = random.randint(int(start_port), int(end_port))
                 # netcat will exit with 0 only if the port is in use,
                 # so a nonzero return value implies it is unused
-                cmd = 'netcat', '0.0.0.0', port, '-w', '1'
+                cmd = 'nc', '0.0.0.0', port, '-w', '1'
                 try:
                     stdout, stderr = utils.execute(*cmd, process_input='')
                 except exception.ProcessExecutionError:

